packages/http/fetch/src/middlewares/parametersNameDecodingHandler.ts

```diff
@@ -0,0 +1,53 @@
+/**
+ * -------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.
+ * See License in the project root for license information.
+ * -------------------------------------------------------------------------------------------
+ */
+
+import { RequestOption } from "@microsoft/kiota-abstractions";
+
+import { Middleware } from "./middleware";
+import { ParametersNameDecodingHandlerOptions, ParametersNameDecodingHandlerOptionsKey } from "./options/parametersNameDecodingOptions";
+
+/**
+ * @module ParametersNameDecodingHandler
+ */
+
+export class ParametersNameDecodingHandler implements Middleware {
+	/**
+	 * @public
+	 * @constructor
+	 * To create an instance of ParametersNameDecodingHandler
+	 * @param {ParametersNameDecodingHandlerOptions} [options = new ParametersNameDecodingHandlerOptions()] - The parameters name decoding handler options value
+	 */
+	public constructor(private options: ParametersNameDecodingHandlerOptions = new ParametersNameDecodingHandlerOptions()) {
+		if (!options) {
+			throw new Error("The options parameter is required.");
+		}
+	}
+	/** @inheritdoc */
+	next: Middleware | undefined;
+	/**
+	 * @public
+	 * @async
+	 * To execute the current middleware
+	 * @param {string} url - The url to be fetched
+	 * @param {FetchRequestInit} requestInit - The request init object
+	 * @param {Record<string, RequestOption>} [requestOptions] - The request options
+	 * @returns A Promise that resolves to nothing
+	 */
+	public execute(url: string, requestInit: RequestInit, requestOptions?: Record<string, RequestOption>): Promise<Response> {
+		let currentOptions = this.options;
+		if (requestOptions && requestOptions[ParametersNameDecodingHandlerOptionsKey]) {
+			currentOptions = requestOptions[ParametersNameDecodingHandlerOptionsKey] as ParametersNameDecodingHandlerOptions;
+		}
+		let updatedUrl = url;
+		if (currentOptions && currentOptions.enable && url.indexOf("%") > -1 && currentOptions.charactersToDecode && currentOptions.charactersToDecode.length > 0) {
```

Comment:
I am familiar with some cases where the Graph API expects the some characters to be double encoded. 

Following is an example: 

https://github.com/microsoftgraph/microsoft-graph-toolkit/pull/1197/files#diff-7d9ac68ab30b3e76b018d22a3f46aa743444fce208ce9f461dcc8b009e275aa9R231

Would this avoid decoding or manipulating the double encoded values which are expected by Graph in some cases?

Reply:
yes, it might decode values as implemented today. However we're only decoding `~ . - $`, and the caller can always disable the middleware, or change the set of characters, either all together or on a per request base.
We've had a long discussion about these cases with @darrelmiller before deciding on an implementation. Effectively the service should never expect a double encoded value. single encoding is fine though.

Reply:
Can you please add a document in this repo something like encoding-decoding.md mentioning this features 
and also create an issue in Graph JS repos to test this feature and mark the issue as test.

Reply:
done https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/753